### PR TITLE
fix(hcu): preserve dense AITER W16A16 dispatch

### PR DESCRIPTION
**Pull request by alexanderbin123**
## Summary

- port the v0.21 HCU W16A16 AITER dispatch preference to the v0.25.1 plugin
- keep the v0.25 `AiterExperts -> rocm_aiter_fused_experts -> rocm_aiter_ops.fused_moe` call chain and call `fused_experts_asm_impl` directly
- patch the completed `vllm.model_executor.layers.fused_moe` package export as well as `layer.FusedMoE`, so real Qwen model imports construct the HCU unquantized method
- bypass generic ROCm MoE padding for AITER W16A16 ASM weights because padding produces strided views while the ASM kernel consumes a dense layout
- preserve Triton and non-AITER upstream behavior, quantized/scaled fallback, and v0.25 ABI validation

## Root cause

Layer-0 A/B capture showed identical hidden states, router logits, top-k weights/IDs, and shared-expert output. The first divergence was the routed-expert output.

Using the real Qwen3.5 layer-0 checkpoint and identical routing inputs:

- dense AITER ASM vs Triton: cosine `0.9999935`, max abs error `4.88e-4`
- ROCm-padded strided view vs Triton: cosine `0.00046`
- dense output norm `1.6069`; padded output norm `0.2516`

The original layer callback also patched only `...fused_moe.layer.FusedMoE`; Qwen imports the cached package-root export, so the real server retained the upstream method and its ROCm padding.

## Verification

- TDD RED: both direct and shuffled weight-preparation cases observed ROCm padded views
- TDD GREEN: AITER ASM preparation keeps dense weights in both modes
- TDD RED: package-root `FusedMoE` remained the original factory
- TDD GREEN: completed package export and layer definition are atomically synchronized
- full plugin suite: `665 passed, 14 warnings`
- `python -m compileall -q vllm_hcu tests`: pass
- `git diff --check`: pass

Hardware, `/models/Qwen3.5-35B-A3B`, HND/bhsd, graph enabled, thinking disabled, temperature 0:

- V2 AITER direct (`shuffle=0`, config=0): smoke `Moon`, IDs `[73418, 248046]`; HumanEval `1/1`
- V2 AITER target (`shuffle=1`, config=1): smoke `Moon`, same IDs; HumanEval-32 `27/32`, pass@1 `0.8438`; 32 predictions and 32 reviews
- V1 AITER target (`shuffle=1`, config=1): smoke `Moon`, same IDs; HumanEval `1/1`
- post-fix V2 Triton: smoke `Moon`, same IDs; HumanEval `1/1`

The AITER runs used explicit `--moe-backend aiter` with both global `VLLM_ROCM_USE_AITER` switches unset/disabled, confirming those environment variables are not required for explicit backend selection.

## Cleanup

All owned server process groups exited cleanly; port 18035 is free and all eight HCUs returned to idle memory.

### CODE REVIEW

**Comment by alexanderbin123** [comment]
## Qwen3.5-35B-A3B：AITER / Triton MoE 对照验证命令

验证配置：物理卡 1、Model Runner V2、HND（FA layout `bhsd`）、Graph 开启、thinking 关闭、`temperature=0`、EvalScope HumanEval 32 条、batch size 1、单条最大输出 4096 tokens。

> Graph 通过不添加 `--enforce-eager` 保持开启。显式指定 `--moe-backend aiter` 时，本次验证没有设置 `VLLM_ROCM_USE_AITER=1` 和 `VLLM_ROCM_USE_AITER_MOE=1`。

### 公共环境

```bash
cd /models/.worktrees/vllm-plugin-das-aiter-w16

unset VLLM_PLUGINS
unset VLLM_ROCM_USE_AITER
unset VLLM_ROCM_USE_AITER_MOE

export HIP_VISIBLE_DEVICES=1
export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_KV_CACHE_LAYOUT=HND
export VLLM_HCU_USE_FLASH_ATTN_UNIFIED=1
export VLLM_HCU_USE_AITER_MOE_CONFIG=1
export VLLM_HCU_USE_AITER_W16A16_MOE_SHUFFLE=1
export VLLM_WORKER_MULTIPROC_METHOD=spawn
export PYTHONPATH=/models/.worktrees/vllm-plugin-das-aiter-w16:${PYTHONPATH}
export NO_PROXY=127.0.0.1,localhost
export no_proxy=127.0.0.1,localhost
```

### AITER MoE 服务端

```bash
vllm serve /models/Qwen3.5-35B-A3B \
  --served-model-name qwen35-a3b-v2-hnd-aiter \
  --trust-remote-code \
  --tensor-parallel-size 1 \
  --moe-backend aiter \
  --gpu-memory-utilization 0.8 \
  --max-model-len 8192 \
  --port 18035 \
  --default-chat-template-kwargs '{"enable_thinking":false}'
```

### AITER EvalScope 客户端

```bash
/tmp/vllm-hcu-acceptance/glm52-pcp-task8-20260819T2121Z/evalscope-venv-virtualenv/bin/evalscope eval \
  --model qwen35-a3b-v2-hnd-aiter \
  --api-url http://127.0.0.1:18035/v1 \
  --api-key EMPTY \
  --eval-type openai_api \
  --generation-config '{"temperature":0,"do_sample":false,"max_tokens":4096,"extra_body":{"chat_template_kwargs":{"enable_thinking":false}}}' \
  --stream \
  --eval-batch-size 1 \
  --timeout 7200 \
  --limit 32 \
  --datasets humaneval \
  --dataset-args '{"humaneval":{}}' \
  --work-dir /tmp/qwen35-a3b-aiter-v2-hnd-graph-he32 \
  --no-timestamp
```

### Triton MoE 服务端

复用上面的公共环境，然后启动：

```bash
vllm serve /models/Qwen3.5-35B-A3B \
  --served-model-name qwen35-a3b-v2-hnd-triton \
  --trust-remote-code \
  --tensor-parallel-size 1 \
  --moe-backend triton \
  --gpu-memory-utilization 0.8 \
  --max-model-len 8192 \
  --port 18035 \
  --default-chat-template-kwargs '{"enable_thinking":false}'
```

### Triton EvalScope 客户端

```bash
/tmp/vllm-hcu-acceptance/glm52-pcp-task8-20260819T2121Z/evalscope-venv-virtualenv/bin/evalscope eval \
  --model qwen35-a3b-v2-hnd-triton \
  --api-url http://127.0.0.1:18035/v1 \
  --api-key EMPTY \
  --eval-type openai_api \
  --generation-config '{"temperature":0,"do_sample":false,"max_tokens":4096,"extra_body":{"chat_template_kwargs":{"enable_thinking":false}}}' \
  --stream \
  --eval-batch-size 1 \
  --timeout 7200 \
  --limit 32 \
  --datasets humaneval \
  --dataset-args '{"humaneval":{}}' \
  --work-dir /tmp/qwen35-a3b-triton-v2-hnd-graph-he32 \
  --no-timestamp
```

### 对照结果

| MoE backend | mean_acc / pass@1 | 通过数 |
|---|---:|---:|
| AITER | 0.8438 | 27/32 |
| Triton | 0.8438 | 27/32 |

两者聚合精度持平。逐题并非完全相同：生成文本完全一致 19/32；AITER 失败题为 `4、10、11、19、20`，Triton 失败题为 `4、9、18、19、20`。两边输出均未发现乱码；Triton 的 HumanEval/19 出现重复生成并达到 4096-token 上限。

**Review by asr-sheep1 — APPROVED** [review]
